specs/experimental/cannon-fault-proof-vm-mt.md

```diff
@@ -0,0 +1,511 @@
+> causing the FPVM to resume normal execution.
+
+### Exited Threads
+
+When the VM encounters an active thread that has exited, it is popped from the active thread stack, removing it from
+the VM state.
+
+### Waiting Threads
+
+Threads enter a waiting state when a futex wait syscall is successfully executed, setting the thread's
+`futexAddr`, `futexVal`, and `futexTimeout` fields according to the futex syscall arguments.
+
+During normal execution, when the active thread is in a waiting state (its `futexAddr` is not `0xFFFFFFFF`), the VM
+checks if it can be woken up.
+
+A waiting thread will be woken up if:
+
+- The current `step` (after incrementing) is greater than `futexTimeoutStep`
+- The memory value at `futexAddr` is no longer equal to `futexVal`
+
+The VM will wake such a thread by resetting its futex fields:
+
+- `futexAddr` = `0xFFFFFFFF`
+- `futexVal` = 0
+- `futexTimeoutStep` = 0
+
+If the current thread is waiting and cannot be woken, it is preempted.
+
+### Voluntary Preemption
+
+In addition to the futex wait syscall (see ["Waiting Threads"](#waiting-threads)), there are a few other syscalls that
+will cause a thread to be "voluntarily" preempted: `sched_yield`, `nanosleep`.
+
+### Forced Preemption
+
+To avoid thread starvation (for example where a thread hogs resources by never executing a sleep, yield, or wait),
+the FPVM will force a context switch if the active thread has been executing too long.
+
+For each step executed on a particular thread, the state field `stepsSinceLastContextSwitch` is incremented.
+When a thread is preempted, `StepsSinceLastContextSwitch` is reset to 0.
+If `StepsSinceLastContextSwitch` reaches a maximum value (`SchedQuantum` = 100_000),
+the FPVM preempts the active thread.
+
+## FPVM State
+
+### State
+
+The FPVM is a state transition function that operates on a state object consisting of the following fields:
+
+1. `memRoot` - A `bytes32` value representing the merkle root of VM memory.
+1. `preimageKey` - `bytes32` value of the last requested pre-image key.
+1. `preimageOffset` - The 32-bit value of the last requested pre-image offset.
+1. `heap` - 32-bit base address of the most recent memory allocation via mmap.
+1. `llReservationActive` - 8-bit boolean indicator of whether a memory reservation,
+   which is reserved via a Load Linked Word (`ll`) instruction, is active.
+1. `llAddress` - 32-bit address of the currently active memory reservation if one exists.
+1. `llOwnerThread` - 32-bit id of the thread that initiated the current memory reservation if one exists.
+1. `exitCode` - 8-bit exit code.
+1. `exited` - 8-bit boolean valuel indicating whether the VM has exited.
+1. `step` - 64-bit step counter.
+1. `stepsSinceLastContextSwitch` - 64-bit step counter that tracks the number of steps executed on the current
+   thread since the last [context switch](https://en.wikipedia.org/wiki/Context_switch).
```

Comment:
Thinking we should rename this to `stepsSinceLastPreemption` since preemption is identical to context switch here. Or at least we should use "preemption" to describe this field.

Reply:
Yeah - good point.  We should clean this up.

Reply:
Added a ticket [here](https://github.com/ethereum-optimism/optimism/issues/12147) just to make sure we don't forget.  But I can probably push that change up real quick.